Implemented Use Case Interactor with Updated Changes to Dialogue Object

ZZBP opened the pull request:
Changes to Dialogue: 

    1. Implemented getters for confidence score and step suggestions.
    2. Changed confidence score to a list of double and step suggestions to a list of strings since we might be returning multiple step suggestions when the confidence score is low.
    3. Changed the return types for confidenceScore and stepSuggestion setters. 

Implementation of TranscriptAnalysisInteractor:
    1. Renamed the interactor to `TranscriptAnalysisInteractor` from StepSuggestionInteractor.
    2. The method in the interactor is now renamed to `analyzeTranscript`, which takes in the parsed transcript and returns the analyzed transcript. The analyzed transcript contains Dialogue objects that have updated step suggestions and confidence scores.
    3. A summary structure of the use case interaction: 

- Generates all the individual step entities. 
- Create the StepManager and SuggestionManager
- In the analyzeTranscrip, Suggestion Manger is called to output step suggestions mapping to their confidence scores. The (As a side note: Suggestion Manger calls StepManager to get the calculated confidence score)
- Lastly, the analyzeTranscrip updates the `confidenceScore` and `stepSuggestion` attributes in Dialogue and returns `transcript`, which is now the analyzed transcript with the updated Dialogue objects.

ZZBP committed:
Updated the return type of confidenceScore in Dialogue;
Implemented StepSuggestionInteractor for our second use case partially. Still need to implement the part for accessing parsed transcript and outputing the suggestions.

ZZBP committed:
Pulled from the origin and updated the interactor.

ZZBP committed:
Dialogue object changes:
1. Implemented getters for confidence score and step suggestion.
2. Changed confidence score to a list of double and step suggestion to a list of strings since we might be returning multiple step suggestions when the confidence score is low.

Notes for TranscriptService:
Discuss the posibility of using builder for downloadTranscript.

ZZBP committed:
Changes to the use case interactor:
1. Renamed the interactor to TranscriptAnalysisInteractor from StepSuggestionInteractor.
2. The main method in the interactor is now renamed as analyzeTranscript which takes in the parsed transcript and returns the analyzed transcript. The analyzed transcript contains Dialogue objects that have updated step suggestion and confidence score.

ZZBP committed:
Added CardStep to TranscriptAnalysisInteractor

Lemiex reviewed (approved):
Couldn't find a thing to fix. Great job Zoey!

ZZBP committed:
Created input boundary for the use case interactor.

auroraborie reviewed (changes requested):
Hey Zoey, 

Good work! I did notice that in accordance with a clean architecture design, the use case interactor should implement an input boundary interface, and accept the input data as a parameter. So before we merge this, could you first have UsecaseInteractor implement a Input Boundary interface? This is just so we do not depend on concretions but abstractions. 

Thanks for the PR!

ZZBP commented:
Hi Aurora! Amazing point! It is definitely essential for us to have the input boundary interface which the use case interaction will implement! Thank you so much for the insight! I have added the input boundary and updated the use case interactor!

ZZBP committed:
Added the checking if the chatbot's message in the Dialogue object is empty in the TranscriptAnalysisInteractor.

auroraborie reviewed (approved):
Hey Zoey this addresses the new features we talked about, of returning the top three choices if none of the talksteps have a confidence score above 70% so good work! I have a PR that will change the name of StepManager and SuggestionManager to a ConfidenceScoreCalculator and ConfidenceScoreOptimizer so you will have to update that after. But I will approve this now.

nm-le reviewed (changes requested):
Great job on adhering to the SOLID principles and clean architecture :) I have some questions and suggestions to change:

1. Do we store confidenceScore in the analyzed transcript and display that to users? Right now I can see that we are only storing the suggested talk step in transcript, but not its corresponding confidenceScore. If we do want to store the score then it should be normalized. Also are we suggesting multiple talk steps (ranked) or just one?

2. In light of the upcoming changes to abstractify Dialogue, we should consider changing methods' names, one of which is analyzeTranscript. But let's merge this branch first.

3. In Dialogue.java: stepSuggestion and confidenceScore are now arraylists and thus should be named accordingly to avoid confusion. For example stepSuggestionList. Line 41 is an example of such confusion (this.confidenceScore.add(confidenceScore);)

4. In TranscriptAnalysisInteractor.java: I want to suggest some changes to the docstring (line 43-54) of analyzeTranscript:
Line 50: Dialogue is a back-and-forth interaction, not conversation. Conversation is defined as one session (or a list of Dialogue objects) corresponding to a conversation id.
Line 51: Speech should be changed to Message

I really like how your code is well-documented. Let's merge soon so we can get on to make the changes in Dialogue.

edwardhuahan commented on backend/src/main/java/com/speakfluid/backend/usecases/TranscriptAnalysisInputBoundary.java:
```diff
@@ -0,0 +1,21 @@
+ */
+
+// use case layer
+public interface TranscriptAnalysisInputBoundary {
```
I like the naming and following conventions great work

edwardhuahan commented on backend/src/main/java/com/speakfluid/backend/usecases/TranscriptAnalysisInteractor.java:
```diff
@@ -0,0 +1,92 @@
+package com.speakfluid.backend.usecases;
```
Good idea, I was not a fan of naming a class UsecaseInteractor since it didn't make much sense to me

edwardhuahan reviewed (approved):
I'm glad someone finally got rid of UsecaseInteractor.java

ZZBP committed:
Adjusted documentation in analyzeTranscript TranscriptAnalysisInteractor
